SignalMessaging/attachments/SignalAttachment.swift

```diff
@@ -767,10 +781,137 @@ public class SignalAttachment: NSObject {
     // NOTE: The attachment returned by this method may not be valid.
     //       Check the attachment's error property.
     private class func videoAttachment(dataSource: DataSource?, dataUTI: String) -> SignalAttachment {
-        return newAttachment(dataSource : dataSource,
-                             dataUTI : dataUTI,
-                             validUTISet : videoUTISet,
-                             maxFileSize : kMaxFileSizeVideo)
+        guard let dataSource = dataSource else {
+            let dataSource = DataSourceValue.emptyDataSource()
+            let attachment = SignalAttachment(dataSource:dataSource, dataUTI: dataUTI)
+            attachment.error = .missingData
+            return attachment
+        }
+
+        if !isInputVideoValidOutputVideo(dataSource: dataSource, dataUTI: dataUTI) {
+            owsFail("building video with invalid output, migrate to async API using compressVideoAsMp4")
+        }
+
+        return newAttachment(dataSource: dataSource,
+                             dataUTI: dataUTI,
+                             validUTISet: videoUTISet,
+                             maxFileSize: kMaxFileSizeVideo)
+    }
+
+    public class func copyToVideoTempDir(url fromUrl: URL) throws -> URL {
+        let baseDir = SignalAttachment.videoTempPath.appendingPathComponent(UUID().uuidString, isDirectory: true)
+        OWSFileSystem.ensureDirectoryExists(baseDir.path)
+        let toUrl = baseDir.appendingPathComponent(fromUrl.lastPathComponent)
+
+        Logger.debug("\(self.logTag) moving \(fromUrl) -> \(toUrl)")
+        try FileManager.default.copyItem(at: fromUrl, to: toUrl)
+
+        return toUrl
+    }
+
+    private class var videoTempPath: URL {
+        let videoDir = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("video")
+        OWSFileSystem.ensureDirectoryExists(videoDir.path)
+        return videoDir
+    }
+
+    public class func compressVideoAsMp4(dataSource: DataSource, dataUTI: String) -> (Promise<SignalAttachment>, AVAssetExportSession?) {
```

Comment:
Hmm. This doesn't feel like it belongs in `SignalAttachment`.

Reply:
Does compressAsJpeg?